Add resource_file to batch pool start_task

stuartleeks opened the pull request:
Closes #2823

Add `resource_file` to `start_task` property in `data_source_batch_pool` and `resource_batch_pool`
Add tests
Add docs

tombuildsstuff commented on azurerm/helpers/azure/batch_pool.go:
```diff
@@ -123,6 +123,20 @@ func FlattenBatchPoolStartTask(startTask *batch.StartTask) []interface{} {
+			resourceFile["http_url"] = armResourceFile.HTTPURL
+			resourceFile["blob_prefix"] = armResourceFile.BlobPrefix
+			resourceFile["file_path"] = armResourceFile.FilePath
+			resourceFile["file_mode"] = armResourceFile.FileMode
```
since these are [all pointers](https://github.com/Azure/azure-sdk-for-go/blob/master/services/batch/mgmt/2018-12-01/batch/models.go#L2650-L2661) - can we nil-check around them and then de-reference them (to ensure they're set correctly); for example:

```
if armResourceFile.AutoStorageContainerName != nil {
  resourceFile["auto_storage_container_name"] = *armResourceFile.AutoStorageContainerName
}
if armResourceFile.StorageContainerURL != nil {
  resourceFile["storage_container_url"] = *armResourceFile.StorageContainerURL
}
if armResourceFile.HTTPURL != nil {
  resourceFile["http_url"] = *armResourceFile.HTTPURL
}
if armResourceFile.BlobPrefix != nil {
  resourceFile["blob_prefix"] = *armResourceFile.BlobPrefix
}
if armResourceFile.FilePath != nil {
  resourceFile["file_path"] = *armResourceFile.FilePath
}
if armResourceFile.FileMode != nil {
  resourceFile["file_mode"] = *armResourceFile.FileMode
}
```

tombuildsstuff commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -232,6 +232,41 @@ func resourceArmBatchPool() *schema.Resource {
+									"auto_storage_container_name": {
+										Type:     schema.TypeString,
+										Optional: true,
+										// TODO ValidateFunc - The autoStorageContainerName, storageContainerUrl and httpUrl properties are mutually exclusive and one of them must be specified.
```
this'll need to be done during the Create/Expand function, unfortunately

stuartleeks replied:
Ooops - I did the implementation in Create/Update but forgot to remove the comment!

tombuildsstuff commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -583,3 +626,43 @@ func validateUserIdentity(userIdentity *batch.UserIdentity) error {
 	return nil
 }
+
+func validateCrossFieldRules(pool *batch.Pool) error {
```
in Go private functions are shared across the entire package; as such could we prefix this:

stuartleeks commented:
Thanks @tombuildsstuff - the changes you requested are in the latest pushed commit :-)

katbyte commented on azurerm/helpers/azure/batch_pool.go:
```diff
@@ -195,11 +222,56 @@ func ExpandBatchPoolStartTask(list []interface{}) (*batch.StartTask, error) {
+	for _, resourceFileValueTemp := range resourceFileList {
+		resourceFileValue := resourceFileValueTemp.(map[string]interface{})
+		resourceFile := batch.ResourceFile{}
+		if autoStorageContainerNameValue, ok := resourceFileValue["auto_storage_container_name"]; ok {
```
Minor, and not required to merge but this would be little cleaner:

```go
		if v, ok := resourceFileValue["auto_storage_container_name"].(string); ok && v != ""{
			resourceFile.AutoStorageContainerName = &v
		}
```

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -394,7 +463,7 @@ EOF
     sku       = "16.04.0-LTS"
     version   = "latest"
-  }
+	}
```
Could we fi the indentation here?
```suggestion
  }
```

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -504,6 +573,242 @@ resource "azurerm_batch_pool" "test" {
         elevation_level = "NonAdmin"
         scope           = "Task"
       }
+		}
```
And here?

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -504,6 +573,242 @@ resource "azurerm_batch_pool" "test" {
+        elevation_level = "NonAdmin"
+        scope           = "Task"
+      }
+		}
```
And here

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -504,6 +573,242 @@ resource "azurerm_batch_pool" "test" {
+        elevation_level = "NonAdmin"
+        scope           = "Task"
+      }
+		}
```
And here

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -504,6 +573,242 @@ resource "azurerm_batch_pool" "test" {
+
+    environment = {
+			env = "TEST",
+			bu  = "Research&Dev"
```
And here

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -504,6 +573,242 @@ resource "azurerm_batch_pool" "test" {
+        elevation_level = "NonAdmin"
+        scope           = "Task"
+      }
+		}
```
And here

katbyte commented on azurerm/resource_arm_batch_pool_test.go:
```diff
@@ -504,6 +573,242 @@ resource "azurerm_batch_pool" "test" {
+        elevation_level = "NonAdmin"
+        scope           = "Task"
+      }
+		}
```
And here

katbyte commented on website/docs/d/batch_pool.html.markdown:
```diff
@@ -92,3 +94,19 @@ A `auto_user` block exports the following:
+
+* `auto_storage_container_name` - The storage container name in the auto storage account.
+
+* `blob_prefix` - The blob prefix used when downloading blobs from an Azure Storage container. Only the blobs whose names begin with the specified prefix will be downloaded. If a prefix is not specified, all the files in the container will be downloaded.
```
Could we reword these properties to better describe a datasource?
```suggestion
* `blob_prefix` - The blob prefix used when downloading blobs from an Azure Storage container.
```

katbyte commented on website/docs/d/batch_pool.html.markdown:
```diff
@@ -92,3 +94,19 @@ A `auto_user` block exports the following:
+
+* `blob_prefix` - The blob prefix used when downloading blobs from an Azure Storage container. Only the blobs whose names begin with the specified prefix will be downloaded. If a prefix is not specified, all the files in the container will be downloaded.
+
+* `file_Mode` - The file permission mode attribute in octal format. This property applies only to files being downloaded to Linux compute nodes. It will be ignored if it is specified for a `resource_file` which will be downloaded to a Windows node. If this property is not specified for a Linux node, then a default value of 0770 is applied to the file.
```
```suggestion
* `file_Mode` - The file permission mode attribute in octal format. 
```

katbyte commented on website/docs/d/batch_pool.html.markdown:
```diff
@@ -92,3 +94,19 @@ A `auto_user` block exports the following:
+
+* `file_Mode` - The file permission mode attribute in octal format. This property applies only to files being downloaded to Linux compute nodes. It will be ignored if it is specified for a `resource_file` which will be downloaded to a Windows node. If this property is not specified for a Linux node, then a default value of 0770 is applied to the file.
+
+* `file_path` - The location on the compute node to which to download the file, relative to the task's working directory. If the `http_url` property is specified, the `file_path` is required and describes the path which the file will be downloaded to, including the filename. Otherwise, if the `auto_storage_container_name` or `storage_container_url` property is specified, `file_path` is optional a ...
```
```suggestion
* `file_path` - The location on the compute node to which to download the file, relative to the task's working directory. If the `http_url` property is specified, the `file_path` is required and describes the path which the file will be downloaded to, including the filename. Otherwise, if the `auto_storage_container_name` or `storage_container_url` property is specified. 
```

katbyte commented on website/docs/d/batch_pool.html.markdown:
```diff
@@ -92,3 +94,19 @@ A `auto_user` block exports the following:
+
+* `file_path` - The location on the compute node to which to download the file, relative to the task's working directory. If the `http_url` property is specified, the `file_path` is required and describes the path which the file will be downloaded to, including the filename. Otherwise, if the `auto_storage_container_name` or `storage_container_url` property is specified, `file_path` is optional a ...
+
+* `http_url` - The URL of the file to download. If the URL is Azure Blob Storage, it must be readable using anonymous access; that is, the Batch service does not present any credentials when downloading the blob. There are two ways to get such a URL for a blob in Azure storage: include a Shared Access Signature (SAS) granting read permissions on the blob, or set the ACL for the blob or its contai ...
```
```suggestion
* `http_url` - The URL of the file to download. If the URL is Azure Blob Storage, it must be readable using anonymous access.
```

katbyte commented on website/docs/d/batch_pool.html.markdown:
```diff
@@ -92,3 +94,19 @@ A `auto_user` block exports the following:
+
+* `http_url` - The URL of the file to download. If the URL is Azure Blob Storage, it must be readable using anonymous access; that is, the Batch service does not present any credentials when downloading the blob. There are two ways to get such a URL for a blob in Azure storage: include a Shared Access Signature (SAS) granting read permissions on the blob, or set the ACL for the blob or its contai ...
+
+* `storage_container_url` - The URL of the blob container within Azure Blob Storage. This URL must be readable and listable using anonymous access; that is, the Batch service does not present any credentials when downloading the blob. There are two ways to get such a URL for a blob in Azure storage: include a Shared Access Signature (SAS) granting read and list permissions on the blob, or set the ...
```
```suggestion
* `storage_container_url` - The URL of the blob container within Azure Blob Storage. 
```

stuartleeks commented:
@katbyte / @tombuildsstuff - I've rebased the changes in this PR now that #3097 is merged, and think I've addressed the PR comments that you left previously. Let me know if there are any other changes you'd like to see before this is merged :-)

tombuildsstuff commented on website/docs/d/batch_pool.html.markdown:
```diff
@@ -110,3 +112,19 @@ A `certificate` block exports the following:
+
+* `blob_prefix` - The blob prefix used when downloading blobs from an Azure Storage container.
+
+* `file_Mode` - The file permission mode attribute in octal format. 
```
I think it's worth calling out this is a string here, since octal format could be interpreted as an int too:

```suggestion
* `file_mode` - The file permission mode attribute represented as a string in octal format (e.g. `"0644"`).
```

tombuildsstuff commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -194,6 +196,24 @@ A `certificate` block supports the following:
+
+* `blob_prefix` - (Optional) The blob prefix to use when downloading blobs from an Azure Storage container. Only the blobs whose names begin with the specified prefix will be downloaded. The property is valid only when `auto_storage_container_name` or `storage_container_url` is used. This prefix can be a partial filename or a subdirectory. If a prefix is not specified, all the files in the contai ...
+
+* `file_Mode` - (Optional) The file permission mode attribute in octal format. This property applies only to files being downloaded to Linux compute nodes. It will be ignored if it is specified for a `resource_file` which will be downloaded to a Windows node. If this property is not specified for a Linux node, then a default value of 0770 is applied to the file.
```
I think it's worth calling out this is a string here, since octal format could be interpreted as an int too:


```suggestion
* `file_mode` - (Optional) The file permission mode represented as a string in octal format (e.g. `"0644"`). This property applies only to files being downloaded to Linux compute nodes. It will be ignored if it is specified for a `resource_file` which will be downloaded to a Windows node. If this property is not specified for a Linux node, then a default value of 0770 is applied to the file.
```

stuartleeks commented:
@tombuildsstuff - thanks for updating the docs :-)

tombuildsstuff commented:
Aside from the long running case sensitivity of the resource group name (which I'll address in a separate PR) - the tests pass/LGTM 👍 

<img width="522" alt="Screenshot 2019-04-16 at 13 13 39" src="https://user-images.githubusercontent.com/666005/56205229-74cd4700-6049-11e9-865c-362553013b38.png">

Thanks again @stuartleeks
